### ISSUE TITLE
[Bug] do_bench_cudagraph starts warmup without waiting for the caller stream

### ISSUE THREAD

**Issue by physics31415926** [FlagTree/flagtree#1127]
## Problem

The public [do_bench_cudagraph implementation](https://github.com/flagos-ai/FlagTree/blob/main/python/triton/testing.py) switches to a newly created stream and immediately invokes the warmup:

```python
with torch.cuda.stream(torch.cuda.Stream()):
    # warmup
    fn()
```

There is no dependency on work already queued on the caller stream before this first call. If the benchmark reads inputs produced by that work, or writes storage that an earlier operation still reads, the warmup can race with the caller. Synchronization later in the helper cannot repair a result already corrupted by that race.

Please establish a caller-to-benchmark stream dependency before warmup. The relevant stream creation pattern was verified in the repository's public main-branch source when filing this report.

## Minimal reproduction

The following synthetic example requires only PyTorch and the benchmark helper, with no model or application code. It deliberately queues an earlier read and makes the benchmark write the same buffer. The earlier read should retain the original value.

```python
import torch
from triton.testing import do_bench_cudagraph

# Initialize the helper before the ordering experiment.
scratch = torch.zeros(1024, device="cuda")
do_bench_cudagraph(lambda: scratch.zero_(), rep=1, n_retries=1)

failures = []
for use_default in (False, True):
    for attempt in range(4):
        buf = torch.ones(1024, device="cuda")
        torch.cuda.synchronize()
        caller = (
            torch.cuda.default_stream()
            if use_default
            else torch.cuda.Stream()
        )
        with torch.cuda.stream(caller):
            torch.cuda._sleep(100000000)
            prior_read = buf.clone()
            do_bench_cudagraph(lambda: buf.fill_(2), rep=1, n_retries=1)
        torch.cuda.synchronize()
        actual = prior_read.unique().tolist()
        if actual != [1.0]:
            failures.append((use_default, attempt, actual))

assert not failures, failures
```

Expected: every earlier read contains only `1.0`. A read containing `2.0` exposes the missing ordering.

This is a timing-dependent race. The private `torch.cuda._sleep` API is used only to widen the reproduction window; its availability and timing may vary by backend. A clean trial alone does not prove the streams are correctly ordered.

## Proposed fix

Replace the implicit temporary stream with an explicitly ordered stream:

```python
caller_stream = torch.cuda.current_stream()
benchmark_stream = torch.cuda.Stream()
benchmark_stream.wait_stream(caller_stream)
with torch.cuda.stream(benchmark_stream):
    # Existing warmup and benchmark body.
    fn()
    ...
```

The wait must be established before the first warmup launch. This creates a GPU stream dependency without adding a host-blocking global synchronization.

A regression test should exercise both default and non-default caller streams and check data ordering, rather than only checking that the helper returns a timing value.

**Comment by physics31415926** [comment]
Confirmed on **Hygon DCU**, using its HIP/HCU backend. NVIDIA hardware has not been tested in this report. The use of `torch.cuda` in the reproduction does not imply NVIDIA: the HCU driver itself exposes that interface.

### Exact code locations

The following links are pinned to the FlagTree main commit checked on 2026-09-08, `eb6c2179a1fe069675949f7199bb7ed68e484dad`:

1. **HCU device interface:** [third_party/hcu/backend/driver.py, lines 860–862](https://github.com/flagos-ai/FlagTree/blob/eb6c2179a1fe069675949f7199bb7ed68e484dad/third_party/hcu/backend/driver.py#L860-L862) returns `torch.cuda`.
2. **HCU replay selection:** [python/triton/flagtune/runtime/benchmark_protocol.py, lines 94–106](https://github.com/flagos-ai/FlagTree/blob/eb6c2179a1fe069675949f7199bb7ed68e484dad/python/triton/flagtune/runtime/benchmark_protocol.py#L94-L106) maps `triton.backends.hcu.driver` to `triton_hcu_graph_replay_v1`.
3. **Replay call into the shared helper:** [benchmark_protocol.py, lines 164–173](https://github.com/flagos-ai/FlagTree/blob/eb6c2179a1fe069675949f7199bb7ed68e484dad/python/triton/flagtune/runtime/benchmark_protocol.py#L164-L173) calls `do_bench_cudagraph` without synchronizing the caller stream first.
4. **Missing dependency before warmup:** [python/triton/testing.py, lines 100–126](https://github.com/flagos-ai/FlagTree/blob/eb6c2179a1fe069675949f7199bb7ed68e484dad/python/triton/testing.py#L100-L126). Lines 106–108 create a new stream and immediately call `fn()`. The synchronization at line 126 occurs after warmup and the initial timing calls.

For clarity, the HCU driver's ordinary [get_benchmarker(), lines 890–892](https://github.com/flagos-ai/FlagTree/blob/eb6c2179a1fe069675949f7199bb7ed68e484dad/third_party/hcu/backend/driver.py#L890-L892) returns `do_bench`. The affected path here is the **replay mode selected through benchmark_protocol**, which uses the shared graph helper.

### Hygon verification using the checked source

We extracted the helper and its required Python helper functions from the pinned source and executed them on Hygon in the existing runtime. This checks the latest function we inspected; it is not a claim that the entire latest FlagTree package was rebuilt and integration-tested.

The expanded independent test covered two data dependencies (an earlier caller-stream read followed by a benchmark write, and a caller-stream write followed by a benchmark read), default/non-default caller streams, and four attempts per combination:

| Treatment on Hygon | Failed cases |
| --- | ---: |
| Original helper from the pinned source | 8/16 |
| Add benchmark_stream.wait_stream(caller_stream) before first warmup | 0/16 |
| Synchronize caller stream before invoking the helper | 0/16 |

The reproduction rate is timing-dependent, but the ordering dependency is still missing in this source.

The implementation being inherited from upstream does not remove the need to satisfy this dependency when replay benchmarking is used during online autotuning. If callers guarantee that prior work has completed, the race may not occur. This HCU replay call path does not establish that guarantee.

The recommended fix remains to make the new benchmark stream wait for the caller stream **before its first warmup**. If the intended scope is HCU-only, the same dependency can be established inside an HCU-specific replay helper. This report does not indicate a defect in HCU computation kernels or the GPU driver.
